Add an option to provide path to `uefi-ntfs.img`

tukusejssirs opened the pull request:
- fixes #11

tukusejssirs commented:
I’ve just created this PR. I tried to test it (manually) if it works correctly (it should), however, @Lin-Buo-Ren, plese review and test it thoroughly.

Just a question: should we also remove [L1079](https://github.com/tukusejssirs/WoeUSB/blob/3dd3cd90aca8a3a6b8f8a23b65b06e7d682a50e2/sbin/woeusb#L1079)?

Lin-Buo-Ren commented on sbin/woeusb:
```diff
@@ -316,6 +316,8 @@ print_help(){
     echo -e
     echo -e '    --for-gui'
     echo -e '        ''No longer supported, reserved for compatibility with the wrapper programs'
+    echo -e '    --uefi-ntfs-img, -u'
```
Missing padding blank like between command-line options

Lin-Buo-Ren commented on sbin/woeusb:
```diff
@@ -492,6 +494,23 @@ process_commandline_parameters(){
+            --uefi-ntfs-img \
+            |-u)
+                shift_array parameters
+                uefi_ntfs_img="${parameters[0]}"
```
We require all variables to be declared in the `local` scope whenever possible, use the name reference feature to allow the variable be modified in a function.

tukusejssirs replied:
> use the name reference feature to allow the variable be modified in a function.

Unfortunately, I have no idea what you mean by _name reference feature_.

In order to comply with your _all variables have to be local_ rule I declared the `$uefi_ntfs_img` variable in the `init()` function (line 64 in the following commit).

Lin-Buo-Ren replied:
> Unfortunately, I have no idea what you mean by name reference feature.

Refer:

* [Shell Parameters (Bash Reference Manual)](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameters.html)
* [Bash Builtins (Bash Reference Manual)](https://www.gnu.org/software/bash/manual/html_node/Bash-Builtins.html)
* [Short Detour into Namerefs | What's New in Bash Parameter Expansion | Linux Journal](https://www.linuxjournal.com/content/whats-new-bash-parameter-expansion#short-detour-into-namerefs)

> A variable can be assigned the nameref attribute using the -n option to the `declare` or `local` builtin commands (see [Bash Builtins](https://www.gnu.org/software/bash/manual/html_node/Bash-Builtins.html#Bash-Builtins)) to create a nameref, or a reference to another variable. This allows variables to be manipulated indirectly. Whenever the nameref variable is referenced, assigned to, unset, or has its attributes modified (other than using or changing the nameref attribute itself), the operation is actually performed on the variable specified by the nameref variable’s value. A nameref is commonly used within shell functions to refer to a variable whose name is passed as an argument to the function. For instance, if a variable name is passed to a shell function as its first argument, running
>
>     declare -n ref=$1
>
> inside the function creates a nameref variable ref whose value is the variable name passed as the first argument. References and assignments to ref, and changes to its attributes, are treated as references, assignments, and attribute modifications to the variable whose name was passed as `$1`.

This feature allows a function to modify the value of the caller's variables (in this case the function is `        process_commandline_parameters` and the caller is `init`)  Note that nameref variables should be named with the `_ref` suffix as bash can't handle nameref variable has the same name with the variable it references.

Lin-Buo-Ren reviewed (commented):
> Just a question: should we also remove [L1079](https://github.com/tukusejssirs/WoeUSB/blob/3dd3cd90aca8a3a6b8f8a23b65b06e7d682a50e2/sbin/woeusb#L1079)?

Keep it as WoeUSB still requires internet access after the merge in some cases.

tukusejssirs commented on sbin/woeusb:
```diff
@@ -492,6 +494,23 @@ process_commandline_parameters(){
+                if \
+                    [[ \
+                        ( "${parameters[*]}" =~ '--tgt-fs NTFS' \
+                        || ( -z "$target_filesystem_type_input" && "$target_filesystem_type_input" = 'NTFS') ) \
```
> Please elaborate on what you're trying to accomplish here

Basically, when one sets both `--tgt-fs` to `NTFS` and `-u` to path + UEFI image, I wanted to test if the argument supplied to `-u` is an existent file.

Actually, it checks the following:

- target filesystem:
   - `$target_filesystem_type_input` might be already defined (when `--tgt-fs` comes before `-u`), therefore we need to test that variable:
      1. if it is defined:
         - we need to use `-z`, because otherwise when the variable is not defined, Bash errors that the variable is unbound;
      2. and at the same time if it is set to `NTFS`.
   - When `$target_filesystem_type_input` is not set, `--tgt-fs` or `--target-filesystem` might be defined later on the command-line, therefore we need to check if `${parameters[*]}` contains them followed by ` NTFS`.

- `$uefi_ntfs_img`:
   - If `$uefi_ntfs_img` is set to a value which is non-existent file, WoeUSB should warn the user about this and revert to downloading the image from the Internet.

> runtime parameter verification should be done in the `check_runtime_parameters` function

Well, how exactly would you like me to do it? Currently, that function accepts the following three arguments:

```bash
check_runtime_parameters \
install_mode \
source_media \
target_media \
"${new_file_system_label}"
```

AFAIK, any of those parameters contain neither filesystem type (e.g. `NTFS`) nor the `$uefi_ntfs_img` variable values (obviously, as it is proposed to add to WoeUSB in this PR). Without adding those information into `check_runtime_parameters` function, I don’t see a way to implement it that way.

> This check doesn't match the `--target-filesystem` long-variant option

Oops, I missed that one.

> Please close all parameters with curly braces

Okay.

> The quoting of `'NTFS'` is unnecessary, please drop them

Okay.

tukusejssirs replied:
We might also want to test the supplied file using `file`:

```bash
file /git/others/rufus/res/uefi/uefi-ntfs.img 
/git/others/rufus/res/uefi/uefi-ntfs.img: DOS/MBR boot sector, code offset 0x3c+2, OEM-ID "mkfs.fat", sectors/cluster 4, root entries 512, sectors 1024 (volumes <=32 MB), Media descriptor 0xf8, sectors/FAT 1, sectors/track 32, heads 64, serial number 0xa8da13d, label: "UEFI_NTFS  ", FAT (12 bit)
```

For example, by grepping `DOS/MBR boot sector` and `label: "UEFI_NTFS`.

Lin-Buo-Ren replied:
> Without adding those information into check_runtime_parameters function, I don’t see a way to implement it that way.

Pass it into that function then?  We can always pass new parameters when necessary.

> We might also want to test the supplied file using `file`

I prefer not doing so, at least not in this PR.

tukusejssirs commented:
Changes to the code:

1. I added `$uefi_ntfs_img_download` variable that should contain Boolean string. If set to `false`, it will use the supplied image (if it is a file), else it will download the UEFI NTFS image. By default, it is set to `true` on line 65.

2. Also, I used the name refs where necessary, as you required.

3. I move the `$uefi_ntfs_img` test to the `check_runtime_dependencies()` function, as you required. Note that I have simplified the condition (a bit) using `grep` (line 702), however, I could not find a better solution to check the `-u`/`--uefi-ntfs-img` then using a `for` cycle, because `$uefi_ntfs_img` might contain spaces. I don’t like the `for` cycle here, because it increases the overhead and depends on `seq`. If you know of a better way, let me know please.

tukusejssirs commented:
@Lin-Buo-Ren, is there anything that needs to be resolved (besides rebasing) before merging this PR? :wink:
